File: LovePlay-UIKit-Masonry/LovePlay/Class/Zone/Controller/DiscussListViewController.m
//
//  DiscussListViewController.m
//  LovePlay
//
//  Created by Yuns on 2017/1/29.
//  Copyright © 2017年 Yuns. All rights reserved.
//

#import "DiscussListViewController.h"
#import "DiscussListModel.h"
#import "DiscussImageModel.h"
#import "DiscussListHeaderView.h"
#import "DiscussListCell.h"
#import "DiscussListTopCell.h"
#import "DiscussDetailViewController.h"

@interface DiscussListViewController ()<UITableViewDataSource, UITableViewDelegate>
//UI
@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) DiscussListHeaderView *headerView;
//Data
@property (nonatomic, strong) NSArray *discussListTopDatas;
@property (nonatomic, strong) NSArray *discussListDatas;
@property (nonatomic, assign) NSInteger pageIndex;
@end

@implementation DiscussListViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    [self initParams];
    [self addTableView];
    [self addTableHeaderView];
    [self loadData];
}

- (void)initParams
{
    _pageIndex = 1;
}

- (void)addTableView
{
    [self.view addSubview:self.tableView];
}

- (void)addTableHeaderView
{
    [self.view addSubview:self.headerView];
}

- (void)loadData
{
    [self loadDiscussListData];
}

- (void)loadDiscussListData
{
    NSDictionary *params = @{@"version":@"163",@"module":@"forumdisplay",@"fid":_fid?_fid:@"",@"tpp":@"15",@"charset":@"utf-8",@"page":@(_pageIndex).stringValue};
    [[HttpRequest sharedInstance] GET:DiscussListURL parameters:params success:^(id response) {
        NSLog(@"list-succ : %@", response);
        DiscussListModel *discussListModel = [DiscussListModel modelWithJSON:response[@"Variables"]];
        [self dealWithDiscussList:discussListModel.forum_threadlist];
        
        [self loadDiscussImageData];
    } failure:^(NSError *error) {
        NSLog(@"list-fail : %@", error);
    }];
}

- (void)loadDiscussImageData
{
    NSString *urlString = [NSString stringWithFormat:@"%@/%@", DiscussListImgURL, _fid];
    [[HttpRequest sharedInstance] GET:urlString parameters:nil success:^(id response) {
        NSLog(@"img-suc : %@", response);
        DiscussImageModel *discussImageModel = [DiscussImageModel modelWithJSON:response[@"info"]];
        [self dealWithDiscussHeaderViewWithDiscussImageModel:discussImageModel];
        [_tableView reloadData];
    } failure:^(NSError *error) {
        NSLog(@"img-Err : %@", error);
    }];
}

- (void)dealWithDiscussHeaderViewWithDiscussImageModel:(DiscussImageModel *)imageModel
{
    
    self.headerView.imageModel = imageModel;
    self.tableView.tableHeaderView = self.headerView;
}

- (void)dealWithDiscussList:(NSArray *)discussList
{
    NSMutableArray *discussListTops = [NSMutableArray array];
    NSMutableArray *discussLists = [NSMutableArray array];
    
    for (ForumThread *forum in discussList) {
        if (forum.displayorder == 1) {
            [discussListTops addObject:forum];
        }else {
            [discussLists addObject:forum];
        }
    }
    _discussListDatas = [discussLists copy];
    _discussListTopDatas = [discussListTops copy];
}

#pragma mark - tableView dataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    switch (section) {
        case 0:
            return _discussListTopDatas.count;
            break;
        case 1:
            return _discussListDatas.count;
            break;
            
        default:
            return 0;
            break;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    switch (indexPath.section) {
        case 0:
        {
            ForumThread *forumThread = _discussListTopDatas[indexPath.row];
            DiscussListTopCell *cell = [DiscussListTopCell cellWithTableView:tableView];
            cell.forumThread = forumThread;
            return cell;
        }
            break;
        case 1:
        {
            ForumThread *forumThread = _discussListDatas[indexPath.row];
            DiscussListCell *cell = [DiscussListCell cellWithTableView:tableView];
            cell.forumThread = forumThread;
            return cell;
        }
            break;
        default:
            return nil;
            break;
    }
}



#pragma mark - tableView delegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    switch (indexPath.section) {
        case 0:
        {
            ForumThread *forumThread = _discussListTopDatas[indexPath.row];
            return [_tableView cellHeightForIndexPath:indexPath model:forumThread keyPath:@"forumThread" cellClass:[DiscussListTopCell class] contentViewWidth:_tableView.width];
        }
            break;
        case 1:
        {
            ForumThread *forumThread = _discussListDatas[indexPath.row];
            return [_tableView cellHeightForIndexPath:indexPath model:forumThread keyPath:@"forumThread" cellClass:[DiscussListCell class] contentViewWidth:_tableView.width];
        }
            break;
        default:
            return CGFLOAT_MIN;
            break;
    }

}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    ForumThread *forumThread = indexPath.section == 0 ? _discussListTopDatas[indexPath.row] : _discussListDatas[indexPath.row];
    DiscussDetailViewController *detailViewController = [[DiscussDetailViewController alloc] init];
    detailViewController.tid = forumThread.tid;
    [self.navigationController pushViewController:detailViewController animated:YES];
}

#pragma mark - Getter / Setter
- (UITableView *)tableView
{
    if (!_tableView) {
        UITableView *tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStylePlain];
        tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
        tableView.delegate = self;
        tableView.dataSource = self;
        _tableView = tableView;
    }
    return _tableView;
}

- (DiscussListHeaderView *)headerView
{
    if (!_headerView) {
        DiscussListHeaderView *headerView = [[DiscussListHeaderView alloc] init];
        headerView.width = self.tableView.width;
        headerView.height = 150;
        _headerView = headerView;
    }
    return _headerView;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: LovePlay-Swift3/LovePlay/Class/Zone/Controller/DiscussListViewController.swift
//
//  DiscussListViewController.swift
//  LovePlay
//
//  Created by weiying on 2017/6/6.
//  Copyright © 2017年 yuns. All rights reserved.
//

import UIKit
import Alamofire

class DiscussListViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    var discussListDatas : [DiscussListModel]? = [DiscussListModel]()
    var discussListTopDatas : [DiscussListModel]? = [DiscussListModel]()
    
    fileprivate var _fid : String?
    var fid : String? {
        set {
            _fid = newValue
        }
        
        get {
            return _fid
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.addSubViews()
        self.loadData()
    }
    
    //MARK: - private
    func addSubViews() {
        self.view.addSubview(self.tableView)
    }
    
    func loadData() {
        self.loadListData()
    }
    
    func loadListData() {
        let params : [String : Any] = ["version":"163", "module":"forumdisplay", "fid":_fid!, "tpp":"15", "charset":"utf-8", "page": "1"]
        let urlStr = DiscussListURL
        Alamofire.request(urlStr, method: .get, parameters: params).responseJSON { (response) in
            switch response.result.isSuccess {
            case true :
                print(response.result.value!)
                if let value = response.result.value as? NSDictionary {
                    let variables = value["Variables"] as? NSDictionary
                    let forum_threadlist = variables?["forum_threadlist"] as? [NSDictionary]
                    var discussList : [DiscussListModel]? = [DiscussListModel]()
                    for dict in forum_threadlist! {
                        let listModel  = DiscussListModel.deserialize(from: dict)
                        discussList?.append(listModel!)
                    }
                    //区分列表数据
                    self.sectionDiscussList(discussList: discussList!)
                    //刷新列表
                    self.tableView.reloadData()
                }
            case false :
                print(response.result.error!)
            }
        }
    }
    
    func loadImgData() {
        
    }
    
    func sectionDiscussList(discussList : [DiscussListModel]) {
        for listModel in discussList {
            if listModel.displayorder == 1 {
                self.discussListDatas?.append(listModel)
            }else{
                self.discussListDatas?.append(listModel)
            }
        }
    }
    
    //MARK: - tableView dataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case 0:
            if (self.discussListTopDatas?.isEmpty)! {
                return 0
            }
            return (self.discussListTopDatas?.count)!
        case 1:
            if (self.discussListDatas?.isEmpty)! {
                return 0
            }
            return (self.discussListDatas?.count)!
        default:
            return 0
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        switch indexPath.section {
        case 0:
            let cell : DiscussListTopCell = DiscussListTopCell.cellWithTableView(tableView: tableView)
            return cell
        case 1:
            let cell : DiscussListCell = DiscussListCell.cellWithTableView(tableView: tableView)
            return cell
        default:
            return UITableViewCell()
        }
    }
    
    //MARK: - tableView delegate
    
    //MARK: - setter / getter
    lazy var tableView : UITableView = {
        let tableView : UITableView = UITableView(frame: self.view.bounds, style: .plain)
        tableView.delegate = self
        tableView.dataSource = self
        tableView.rowHeight = UITableViewAutomaticDimension
        tableView.estimatedRowHeight = 60
        return tableView
    }()

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    


}
